receiver/sqlqueryreceiver/internal/database_sql.go

```diff
@@ -0,0 +1,55 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package internal // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/sqlqueryreceiver/internal"
+
+import (
+	"database/sql"
+	"sync"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/sqlquery"
+)
+
+func NewPool(opener sqlquery.SQLOpenerFunc, driver string, dsn string, maxOpenConns int) interface {
+	DB() (*sql.DB, error)
+} {
+	return &sqlPool{
+		DriverName:     driver,
+		DataSourceName: dsn,
+		MaxOpenConns:   maxOpenConns,
+		Opener:         opener,
+	}
+}
+
+type sqlPool struct {
+	DriverName     string
+	DataSourceName string
+	MaxOpenConns   int
+	Opener         sqlquery.SQLOpenerFunc
+
+	mutex sync.Mutex
+	db    *sql.DB
+}
+
+// DB initializes and returns the [sql.DB] pool. It is safe to call concurrently.
+// Depending on the driver, this might also connect to the database backend.
+//
+// This method exists to satisfy [sqlquery.DbProviderFunc], but the way
+// [sqlquery.Scraper] closes [sql.DB] can interfere with other Scrapers.
+func (sp *sqlPool) DB() (*sql.DB, error) {
+	sp.mutex.Lock()
+	defer sp.mutex.Unlock()
```

Comment:
I'm concerned about some edge case where this mutex hangs due to the `sp.Opener` call. Have this been tested with an invalid database configuration, or do you have some other idea for how to test?

Reply:
I understand your concern. You're worried that a slow Open will prevent this receiver's scrapers from scraping. That's how it goes if we want to share the pool among the scrapers.

Are you worried that this might affect startup/configuration? This function is called after config, when the receiver starts.

I can mock an opener that returns an error. 🤔 That would release the mutex and bubble up during Start. A panic here would do the same.

Reply:
It might be alright for now, we can revisit this if we end up running into any issues. 👍